doc/tutorial.rst

```diff
@@ -0,0 +1,317 @@
+between the baseline (``--- expected``) and the actual output (``+++ output``)
+showing the difference, and the testsuite's ``--show-error-output/-E`` option
+displays it, making it easy to quickly spot the difference between the two.
+
+Even though these 3 testcases take very little time to run, most testsuites
+require a lot of CPU time to run to completion. Nowadays, most working stations
+have several cores, so we can spawn one test per core to speedup testsuite
+execution time. ``e3.testsuite`` supports the ``--jobs/-j`` option to achive
+this. This option works the same way it does for the ``make`` program: ``-jN``
+is the default (run at most N testcases at a time, default is 1), and ``-j0``
+tells to set N to the number of CPU cores.
+
+
+Test execution control
+----------------------
+
+There is no obvious bug in ``bc`` that this documentation could expect to
+survive for long, so let's stick with this wrong ``multiplication`` testcase
+and pretend that ``bc`` should return ``8``. This is a known bug, and so the
+failure is expected for the time being. This situation occurs a lot in
+software: bugs often take a lot of time to fix, sometimes test failures come
+from bugs in upstream projects, etc.
+
+To keep testsuite reports readable/usable, it is convenient to tag failures
+that are temporarily accepted as ``XFAIL`` rather than ``FAIL``: the former
+is a failure that has been analyzed as acceptable for now, leaving the latter
+for unexpected regressions to investigate. Testcases using a driver that
+inherits from ``ClassicTestDriver`` can do that adding a ``control`` entry in
+their ``test.yaml`` file. To do that, append the following to
+``multiplication/test.yaml``:
+
+.. code-block:: yaml
+
+   control:
+   - [XFAIL, "True", "erroneous multiplication: see bug #1234"]
+
+When present, ``control`` must contain a list of 2- or 3-uplets:
+
+* A command. Here, ``XFAIL`` to state that failure is expected: ``FAIL`` test
+  statuses are turned into ``XFAIL``, and ``PASS`` are turned into ``XPASS``.
+  There are two other commands available: ``NONE`` (the default: regular test
+  execution), and ``SKIP`` (do not execute the testcase and create a ``SKIP``
+  test result).
+* A Python expression as a condition guard, to decide whether the command
+  applies to this testsuite run. Here, it always applies.
+* An optional message to describe why this command is here.
+
+The first command whose guard evaluates to true applies. We can see it in
+action:
+
+.. code-block:: sh
+
+   $ ./testsuite.py -j8
+   INFO     Found 3 tests
+   INFO     XFAIL        multiplication: unexpected output (erroneous multiplication: see bug #1234)
+   INFO     PASS         subtraction
+   INFO     PASS         addition
+   INFO     Summary:
+   _  PASS         2
+   _  XFAIL        1
+
+You can learn more about this specific test control mechanism and even how to
+create your own mechanism in :ref:`api_control`.
+
+
+What's next
+-----------
+
+ref_result:
+
+* ???
```

Comment:
To be updated.

Reply:
To be removed, even! I should have written TODO… I forgot to remove it afterwards